### ISSUE TITLE
Accessibility: Inaccessible/confusing modal dialog when canceling on Add Calendar Block screens

### ISSUE THREAD

**Issue by ndarilek** [loconomics/loconomics#632]
When I add a calendar block, click the button to set a start time, then click Cancel, I think I'm placed in a modal dialog. If this is the case:

 * The modal isn't reviewable with the caret. I wonder if its appearance clashes with the focus-setting code which triggers on navigation to new activities?
 * In addition to Yes and no buttons, there's also a Close button. I'm not sure what benefit the Close button offers over, say, No. It should probably be removed.

See [this resource](https://gdkraus.github.io/accessible-modal-dialog/) for making modals accessible.

**Comment by IagoSRL** [comment]
I was testing this with NVDA in Firefox at Windows 7 (with F7 cursor mode enabled), and this is what I found:

> The modal isn't reviewable with the caret

I have forgot the attribute `role=document` for the element inside the modal container as recommended by Bootstrap examples; adding it makes that (after press Tab on NVDA) the screen reader says 'document' and start reading the content, starting by the Close button, and it allows to use the caret.

> In addition to Yes and no buttons, there's also a Close button. I'm not sure what benefit the Close button offers over, say, No. It should probably be removed.

Does make it consistence with other modals (too it is possible to press Esc key with same result), but I understand what you say, on this case can make sense to remove that. But, after remove it I had again the same problem than before adding `role=document`: the screen reader don't reads the content text, on pressing Tab it just goes to the No button; I had to press Up Arrow to make it go back and read the text.
I tried another thing, adding `tabindex=0` to the container of the text ('modal-body'); on this case pressing Tab after the modal being announced it focus the text content and read it; but there is an strange thing, the screen reader says "document. Clickable 'Are you sure'. Are you sure?". It seems to believe the text is clickable (that's wrong and confusing), and announce it twice. Discarded.

I tried your resource link with the modal example. Strange, but it does not read the visible modal text, only a hidden text (visible only to screen reader) that is linked with an `aria-describedby` attribute, making that the modal reads that after announce the dialog; immediately, the first form field is focused, being announced when finishing reading the `aria-describedby` text, without user requiring to press Tab.

So, for our case I tried linking the dialog content with that attribute. Now is announced after the modal title. Press tab make it goes to the first button ('No') with a side effect: it says 'document. Clickable No button. Button No. Button Yes'.
Removed the `role=document` and it seems 'perfect'. The document is still reachable with keyboard on my case.
@ndarilek I uploaded the code with this changes to the is528 branch, please try when you can and let me know how goes for you and any recommendation if possible.

**Comment by ndarilek** [comment]
I haven't been able to navigate back to this screen--the interface is a bit confusing and I don't quite get how everything is supposed to fit together--but canceling on the Add Client screen gives me the exact same inaccessible behavior. Was this change made on all app dialogs, or on just the one I pointed out?

Note that you may also have to set `aria-hidden="true"` on the rest of the document, and `aria-hidden="false"` on the modal. I did some tests on this a couple years ago, and discovered that screen readers have a hard time correctly representing focus when a modal is open. Making the modal the only item visible to the screen reader helped with that.

**Comment by ndarilek** [comment]
Please disregard my bit about not finding this screen. My browser window wasn't maximized, causing the page to render at a smaller viewport breakpoint. I found the screen by clicking on the various collapsed navigation buttons.

And, yes, this dialog is unchanged. Try the `aria-hidden` trick and see if that helps.

**Comment by IagoSRL** [comment]
OK, I will try. Too I will need to set-up again an Ubuntu to try with Orca, seems to have lot of differences with NVDA results on Windows :-/

**Comment by ndarilek** [comment]
Actually, not sure why I didn't think of this before. Here's a library 
you might wish to add to Loconomics, full of lots of accessibility helpers:

https://allyjs.io/

In particular, look into this API:

https://allyjs.io/api/maintain/hidden.html

I'll bet that if you pass in a selector to the modal dialog, you'll find 
that it solves my issue.

Interesting that it works for you under NVDA. When I did my own 
NVDA-based modal testing a year or so ago, this exact issue with these 
same symptoms tripped me up. Perhaps something changed in NVDA in the 
interim, which is great as this fix is completely unintuitive and took 
even me a while to figure out. :)

**Comment by IagoSRL** [comment]
@ndarilek I just pushed some changes to modals of type 'confirm', I followed your advice for aria-hidden all elements except the modal (learning some detail from the allyjs helped, thanks; ended not using it but creating specific code but helped me).

I set-up a special activity at URL http://localhost:8811/appDebug.html#!/_modals just for testing using modals without need to go to specific use cases; we may discover too problems part of the activity using it rather than the modal itself.

When we have this working fine, I will continue applying the fixes to the other modals.

**Comment by IagoSRL** [comment]
@ndarilek I make changes in all other modals already, using the aria-hidden there too and other markup details. I fixed too a general bug with 'focus', now when closing the modal the focus returns to the button that opened the modal.

But, there is an annoying side effect of the aria-hidden tip, almost with NVDA and Firefox: when clicking the button that opens the modal, the label of the button is read again, just before read the dialog; not sure if happens at other screen readers and how annoying is for users.

**Comment by IagoSRL** [comment]
Forgot to say that I updated the testing '_modals' activity with examples for all the modal types that exists.

**Comment by ndarilek** [comment]
Hey, these are great! Quite usable and almost perfect!

For bonus points, you might want to implement some of the markup and 
tips at 
https://www.marcozehe.de/2015/02/05/advanced-aria-tip-2-accessible-modal-dialogs/. 
In particular, the sections titled "


    The simple Confirmation dialog", where aria-describedby/aria-title
    are shown, and the section "Initial focus" where focus moves to a
    button would be helpful. But those are just nice-to-haves.


I'm not sure about the double-speaking of the button label. That may go 
away if the dialog is correctly labeled as above. But either way, I'd 
hardly call that a major need.

**Comment by IagoSRL** [comment]
Thanks!

The article looks interesting. We can save this as enhancements and review later.

The 'initial focus' thing I already tried it: our text editor modal had that, but I'm having problems with that because the screen reader started saying things that makes no sense (the exact NVDA behavior was to read, read the description 'section section section' and then the label of the input). So I discarded, but can be researched further (and me doing more cross screen readers + systems testing).

**Comment by IagoSRL** [comment]
Opened #657 #658 for the small details commented lately here (initial focus and NVDA double speech of button label), and closing this since the important stuff is done.
